apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(byte[] message) =>
+            PublishAsync(new PubsubMessage
+            {
+                Data = ByteString.CopyFrom(message)
+            });
+
+        /// <summary>
+        /// Retrieve a snapshot of the flow state.
+        /// </summary>
+        /// <returns></returns>
+        public virtual FlowState GetCurrentFlowState() => throw new NotImplementedException();
+
+        /// <summary>
+        /// Shutdown this <see cref="HiPerfPublisher"/>.
+        /// The returned <see cref="Task"/> completes when all queued messages have been published.
+        /// The returned <see cref="Task"/> cancels if the passed <see cref="CancellationToken"/> is cancelled.
+        /// Cancelling the passed <see cref="CancellationToken"/> aborts publishing, and cancels the returned
+        /// <see cref="Task"/> as quickly as possible.
+        /// </summary>
+        /// <param name="hardStopToken">Cancel this <see cref="CancellationToken"/> to abort publishing queued messages.</param>
+        /// <returns>A <see cref="Task"/> that completes when all queued messages have been published; or cancels if the
+        /// passed <see cref="CancellationToken"/> is cancelled.</returns>
+        public virtual Task ShutdownAsync(CancellationToken hardStopToken) => throw new NotImplementedException();
+
+        /// <summary>
+        /// Shutdown this <see cref="HiPerfPublisher"/>.
+        /// The returned <see cref="Task"/> completes when all queued messages have been published.
+        /// The returned <see cref="Task"/> cancels if the passed timeout expires before all messages are published.
+        /// </summary>
+        /// <param name="timeout">After this period, abort publishing queued messages.</param>
+        /// <returns>A <see cref="Task"/> that completes when all queued messages have been published; or cancels if the
+        /// passed timeout expires.</returns>
+        public virtual Task ShutdownAsync(TimeSpan timeout) => ShutdownAsync(new CancellationTokenSource(timeout).Token);
+    }
+
+    /// <summary>
+    /// Implementation of PubSub publisher that is associated with a specific <see cref="TopicName"/>.
+    /// </summary>
+    public sealed class HiPerfPublisherImpl : HiPerfPublisher
+    {
+        private class Batch
+        {
+            public TaskCompletionSource<IList<string>> BatchCompletion { get; } = new TaskCompletionSource<IList<string>>();
+            public List<PubsubMessage> Messages { get; } = new List<PubsubMessage>();
+            public CancellationTokenSource TimerCts { get; } = new CancellationTokenSource();
+            public long ByteCount { get; private set; }
```

Comment:
So this is the byte count of the messages, not necessarily the byte count of the request itself (which will include other bits) right?